GPA layer and NLL loss functiona added

hasan2m opened the pull request:
Unit test passed for both GPA layer and loss function. 

GPA layer: 

![image](https://github.com/user-attachments/assets/eb4390ca-65c5-4ea7-8c1f-7a0f982a3f60)

Loss function:

![image](https://github.com/user-attachments/assets/3c01398d-f7e4-410a-908c-bd613af76208)

hasan2m committed:
GP layer and NLL loss functiona added

sgoldenCS commented on jlab_datascience_toolkit/utils/keras_layers/GP_layer.py:
```diff
@@ -0,0 +1,229 @@
+            initializer=tf.constant_initializer(self.initial_noise_scale),
+            trainable=self.train_noise_scale,
+            dtype=tf.float32,
+            constraint=ClipByValue(1e-6, 1e6),
```
I like the use of a function for the constraint, but would it be possible to continue using the `noise_bounds` parameter here? Otherwise it should be removed from the class definition as I don't believe it's used elsewhere.

hasan2m replied:
Removed noise_bound from the class definition.

sgoldenCS commented on jlab_datascience_toolkit/utils/keras_layers/GP_layer.py:
```diff
@@ -0,0 +1,229 @@
+        )
+
+
+        # self.noise_scale = tf.Variable(
+        #     self.initial_noise_scale,
+        #     dtype=tf.float32,
+        #     trainable=self.train_noise_scale,
+        #     constraint=lambda z: tf.clip_by_value(z, self.noise_bounds[0], self.noise_bounds[1]),
+        #     name='noise_scale'
+        # )
```
This looks to be essentially the same as the new version (other than the `noise_bounds` thing), so feel free to remove the commented section to clean things up.

sgoldenCS commented on jlab_datascience_toolkit/utils/keras_layers/GP_layer.py:
```diff
@@ -0,0 +1,229 @@
+
+    #     batch_size = tf.cast(tf.shape(inputs)[0], tf.float32)
+    #     x = tf.convert_to_tensor(inputs, dtype=self.dtype)
+    #     x = tf.cast(x, tf.float32)
+        
+    #     x = self.length_scale * x
+    #     x = self.rff_map(x)
+    #     x1 = tf.math.cos(x) 
+    #     x2 = tf.math.sin(x) 
+
+    #     ffs = layers.concatenate([x1, x2])
+        
+    #     if self.scale_features:
+    #         ffs = tf.math.sqrt(2.0 / self.n_fourier_features) * ffs
+        
+    #     ffs = tf.math.sqrt(self.constant_scale) * ffs
+    #     output = self.rff_output(ffs)
+        
+    #     if training:
+    #         if self.momentum > 0:
+    #             update_prior_op = (
+    #                 self.momentum * self.prior + (1 - self.momentum) * (tf.transpose(ffs) @ ffs / batch_size)
+    #             )
+    #         else:
+    #             update_prior_op = self.prior + tf.transpose(ffs) @ ffs
+    #         self.prior.assign(update_prior_op)  # Direct assignment
+
+    #         variances = self.calc_variance(ffs)
+    #     else:
+    #         if not self.do_custom_cov_update:
+    #             self.update_cov(self.prior)
+
+    #         variances = self.calc_variance(ffs)
+
+    #     stddevs = tf.math.sqrt(variances)
+    #     out = [output, stddevs[:, None]]
+    #     if return_features:
+    #         out.append(ffs)
+
+    #     return out
```
I don't see any changes here other than the default for the `training` flag, so I think this commented section could be removed too.

schr476 replied:
clean up code.

sgoldenCS reviewed (commented):
I generally think things look fine although I haven't run the unittests myself (maybe someone else could go through that process?). I'm glad you're using pytest and pytest fixtures though! It looks good!

schr476 commented on jlab_datascience_toolkit/utils/keras_layers/GP_layer.py:
GaussianProcessLayer --> We should rename since this is RFF approximation and not GP

hasan2m replied:
Name changed to GaussianProcessApproxiamtionLayer.

schr476 replied:
remove this from utils.

schr476 commented on jlab_datascience_toolkit/utils/keras_layers/GP_layer.py:
please change file name --> gpa_layer.py

schr476 commented on jlab_datascience_toolkit/utils/keras_losses/nll_loss.py:
```diff
@@ -0,0 +1,26 @@
+        tf.Tensor: The computed NLL loss (scalar).
+    """
+    
+    sigma_star = tf.square(std) + noise_scale + 1e-5  # Adding a small constant for numerical stability
```
is the noise scale additive?

hasan2m replied:
Removing the noise_scale from here because we are accounting the noise scale during the variance calculation within the gpa_layer class.

schr476 commented on utests/utest_gp_layer.py:
```diff
@@ -0,0 +1,60 @@
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras import Model, Input
+from jlab_datascience_toolkit.utils.keras_layers.GP_layer import GaussianProcessLayer
```
update with request change above.
Please don't call it a GaussianProcessLayer.

schr476 commented on utests/utest_gp_layer.py:
```diff
@@ -0,0 +1,60 @@
+
+
+def test_forward_pass(random_input):
+    layer = GaussianProcessLayer()
```
Please don't call it GaussianProcessLayer

hasan2m committed:
addressed the comments on PR

hasan2m commented:
Addressed the comments and reran the unit test. 

![image](https://github.com/user-attachments/assets/e2209e2a-239a-46bb-a371-9c9ea26e5bee)

![image](https://github.com/user-attachments/assets/8132c895-fa36-40d9-8622-f5c4bcce229e)

Kishanrajput commented:
@hasan2m - could you please move the layer and loss out from utils by following below directory structure
Toolkit
 - keras (dir)
   - layers
     - layer_v0.py (implementation)  
   - losses (dir)
     - loss_v0.py (implementation)

schr476 commented on jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py:
```diff
@@ -0,0 +1,174 @@
+
+        if training:
+            update_prior_op = (
+                self.momentum * self.prior + (1 - self.momentum) * (tf.transpose(ffs) @ ffs / batch_size)
```
check of the size is correct: (batch,batch)

schr476 commented on jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py:
```diff
@@ -0,0 +1,174 @@
+        self.eigvals.assign(eigvals)
+        self.eigvecs.assign(eigvecs)
+
+    def calc_variance(self, ffs):
```
need to understand this better. this should be K_xx^{-1}

hasan2m replied:
Inverse of K_xx is determined using eigenvalue and eigenvectors.

schr476 commented on jlab_datascience_toolkit/utils/keras_layers/gpa_layer.py:
```diff
@@ -0,0 +1,174 @@
+            name='rff_map'
+        )
+        
+       self.rff_output = layers.Dense(self.n_out, use_bias=False, name='GP_mean_pred')
```
What is this here?

hasan2m replied:
Output layer that provides the mean prediction using the fourier features created by rff_map.

schr476 commented on jlab_datascience_toolkit/utils/keras_losses/nll_loss.py:
```diff
@@ -0,0 +1,25 @@
+        tf.Tensor: The computed NLL loss (scalar).
+    """
+    
+    sigma_star = tf.square(std) + 1e-5  # Adding a small constant for numerical stability
```
sigma_star should be sigma2_star

hasan2m replied:
corrected.

hasan2m committed:
New directory

hasan2m committed:
Minor change

schr476 commented:
@hasan2m what is the status of the requested changes?

sgoldenCS commented on jlab_datascience_toolkit/keras/layers/gpa_layer_v0.py:
```diff
@@ -0,0 +1,175 @@
+        return result
+
+    
+    def update_cov(self, prior):
+        eigvals, eigvecs = tf.linalg.eigh(prior)  #EigenDecomposition
+        eigvals = tf.where(eigvals > 0, eigvals, tf.zeros_like(eigvals))
+
+        self.eigvals.assign(eigvals)
+        self.eigvecs.assign(eigvecs)
```
I wonder if this should be an internal version (`_update_cov(self, prior)` or something similar) that can be called with any prior, and implement the `update_cov` function without the extra input parameter like this:
```python
def update_cov(self):
    self._update_cov(self.prior)
```
That could make it easier for a user to update the variance calculations outside of the model. Going along with this, it might also make sense to just remove the `update_cov()` call on line 146, and just rely on the user to properly update the prior when desired. At the moment, if `do_custom_cov_update == False`, the code will be slow for every non-training/validation call, which probably isn't desirable.

Kishanrajput reviewed (changes requested):
When a model that uses this layer is saved and loaded back - it does not save/load prior matrix which makes the saved models not useful. Please add prior saving and loading.

Kishanrajput commented:
@hasan2m - please have a look at the comments and address/comment on them.

Kishanrajput commented:
Print statement for the RFF shape should be removed
